d2/src/test/java/com/linkedin/d2/balancer/util/WarmUpLoadBalancerTest.java

```diff
@@ -316,22 +358,124 @@ public void testHitTimeout() throws URISyntaxException, InterruptedException, Ex
       "Expected # of requests between " + expectedRequests + " +/-" + deviation + ", found:" + requestCount.get());
   }
 
+  @DataProvider // to test dual read modes under which the specific type of warmup load balancer should do warmup
+  public Object[][] modesToWarmUpDataProvider()
+  {
+    return new Object[][]
+        {// @params: {dual read mode, isIndis}
+            {NEW_LB_ONLY, true},
+            {OLD_LB_ONLY, false},
+            // under dual read mode, both INDIS and ZK warmup should do warmup
+            {DUAL_READ, true},
+            {DUAL_READ, false}
+        };
+  }
 
-  @BeforeMethod
-  public void createTempdir() throws IOException
+  @Test(dataProvider = "modesToWarmUpDataProvider")
+  public void testSuccessWithDualRead(DualReadModeProvider.DualReadMode mode, Boolean isIndis)
+      throws InterruptedException, ExecutionException, TimeoutException
   {
-    _tmpdir = LoadBalancerUtil.createTempDirectory("d2FileStore");
-    _FSBasedDownstreamServicesFetcher = new FSBasedDownstreamServicesFetcher(_tmpdir.getAbsolutePath(), MY_SERVICES_FS);
+    int warmUpTimeout = 4;
+    createDefaultServicesIniFiles();
+    setDualReadMode(mode);
+
+    // 3 dual read fetches take 1.5s, 3 warmups take at most 3 * (500 +/- 10) ms. Total at most is 3030 ms.
+    TestLoadBalancer balancer = new TestLoadBalancer(500, 500);
```

Comment:
Is it possible to keep this in single digit millis? Tests taking long to execute add up quickly to result in a slower build.

Reply:
I thought about this too. But the warmup timeout config takes in seconds, so to test the timeout case, the dual read fetching got be more than at least 1s.

Reply:
In that case, it might be best to add another constructor with `Supplier<Long> timeSupplier` as a param, which is set to `() -> SystemClock.instance().currentTimeMillis()` for the existing constructors. This should make the tests more reliable while avoiding the penalty of wasted time.

Reply:
This. Please make any time source pluggable through a constructor. Otherwise you will always have issues calling into some static clock with tests.

Reply:
but there will be multiple calls to get the time (2 critical ones --- one in fetching dual read, one in actual warmup work. And many other ones for logs) which need different returns. It will be hard to code what time to return for which call.

Reply:
Time always moves forward (I hope) so would a supplied that returns a monotonically increasing value work?

Reply:
> a supplied that returns a monotonically increasing value

Could you list a sample snippet? just to make sure I understand what it means.

Reply:
Something like this:

```java
new Supplier<Long>() {
  long time = 0;

  @Override
  public Long get() {
    return time += 100;
  }
}
```

Every time the supplier is called it will "move time forward" by 100 milliseconds. Would something like this work for the tests?

Reply:
updated. Thanks!

Reply:
Thanks, keep the tests nice and fast.